File: packages/cli/test/merge.test.ts
import { describe, expect, it } from 'vitest'
import * as Y from 'yjs'
import { mergeExternalEdit } from '../src/merge.ts'

function textOf(content: string): Y.Text {
  const doc = new Y.Doc()
  const text = doc.getText('content')
  text.insert(0, content)
  return text
}

describe('mergeExternalEdit', () => {
  it('applies the edit when nothing else changed', () => {
    const text = textOf('# title\n\nbody\n')
    mergeExternalEdit(text, '# title\n\nbody\n', '# Title\n\nbody\nmore\n', 'file')
    expect(text.toString()).toBe('# Title\n\nbody\nmore\n')
  })

  it('keeps remote edits made elsewhere in the document', () => {
    const base = 'one\ntwo\nthree\n'
    const text = textOf(base)
    text.insert(0, 'zero\n') // remote edit, not yet on disk
    mergeExternalEdit(text, base, 'one\ntwo\nthree\nfour\n', 'file')
    expect(text.toString()).toBe('zero\none\ntwo\nthree\nfour\n')
  })

  it('keeps both sides when they touch the same line', () => {
    const base = 'hello world\n'
    const text = textOf(base)
    text.insert(5, ',') // remote: "hello, world"
    mergeExternalEdit(text, base, 'hello world!\n', 'file')
    expect(text.toString()).toBe('hello, world!\n')
  })

  it('lets an external deletion remove the region it replaced', () => {
    const base = 'a\nb\nc\n'
    const text = textOf(base)
    text.insert(0, 'x') // remote
    mergeExternalEdit(text, base, 'a\nc\n', 'file')
    expect(text.toString()).toBe('xa\nc\n')
  })

  it('tags the transaction with the given origin', () => {
    const text = textOf('a')
    const origins: unknown[] = []
    text.doc?.on('update', (_u: Uint8Array, origin: unknown) => origins.push(origin))
    mergeExternalEdit(text, 'a', 'ab', 'file')
    expect(origins).toEqual(['file'])
  })
})

File: internal/merge/merge_test.go
package merge

import (
	"testing"

	"github.com/reearth/ygo/crdt"
)

func textOf(content string) (*crdt.Doc, *crdt.YText) {
	doc := crdt.New()
	text := doc.GetText("content")
	doc.Transact(func(txn *crdt.Transaction) { text.Insert(txn, 0, content, nil) })
	return doc, text
}

// remote simulates an edit from another peer that is not on disk yet.
func remote(doc *crdt.Doc, text *crdt.YText, index int, s string) {
	doc.Transact(func(txn *crdt.Transaction) { text.Insert(txn, index, s, nil) })
}

func TestExternalEdit(t *testing.T) {
	tests := []struct {
		name   string
		base   string
		remote func(*crdt.Doc, *crdt.YText)
		next   string
		want   string
	}{
		{
			name: "applies the edit when nothing else changed",
			base: "# title\n\nbody\n",
			next: "# Title\n\nbody\nmore\n",
			want: "# Title\n\nbody\nmore\n",
		},
		{
			name:   "keeps remote edits made elsewhere in the document",
			base:   "one\ntwo\nthree\n",
			remote: func(d *crdt.Doc, x *crdt.YText) { remote(d, x, 0, "zero\n") },
			next:   "one\ntwo\nthree\nfour\n",
			want:   "zero\none\ntwo\nthree\nfour\n",
		},
		{
			name:   "keeps both sides when they touch the same line",
			base:   "hello world\n",
			remote: func(d *crdt.Doc, x *crdt.YText) { remote(d, x, 5, ",") },
			next:   "hello world!\n",
			want:   "hello, world!\n",
		},
		{
			name:   "lets an external deletion remove the region it replaced",
			base:   "a\nb\nc\n",
			remote: func(d *crdt.Doc, x *crdt.YText) { remote(d, x, 0, "x") },
			next:   "a\nc\n",
			want:   "xa\nc\n",
		},
		{
			name:   "counts offsets in UTF-16 units around surrogate pairs",
			base:   "🌏 こんにちは\n",
			remote: func(d *crdt.Doc, x *crdt.YText) { remote(d, x, 2, "🐹") },
			next:   "🌏 こんばんは\n😀\n",
			want:   "🌏🐹 こんばんは\n😀\n",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			doc, text := textOf(tt.base)
			if tt.remote != nil {
				tt.remote(doc, text)
			}
			ExternalEdit(doc, text, tt.base, tt.next, "file")
			if got := text.ToString(); got != tt.want {
				t.Fatalf("got %q, want %q", got, tt.want)
			}
		})
	}
}

func TestExternalEditTagsOrigin(t *testing.T) {
	doc, text := textOf("a")
	var origins []any
	doc.OnUpdate(func(_ []byte, origin any) { origins = append(origins, origin) })
	ExternalEdit(doc, text, "a", "ab", "file")
	ExternalEdit(doc, text, "ab", "ab", "file") // no change, no transaction
	if len(origins) != 1 || origins[0] != "file" {
		t.Fatalf("origins = %v", origins)
	}
}
